appserver/web/web-glue/src/main/java/com/sun/web/server/J2EEInstanceListener.java

```diff
@@ -203,7 +203,7 @@ else if (baseHttpServletRequest instanceof RequestFacade) {
 
                 if (principal != null && principal == basePrincipal && principal.getClass().getName().equals(WEB_PRINCIPAL_CLASS)) {
                     securityContext.setSecurityContextWithPrincipal(principal);
-                } else if (principal != basePrincipal) {
+                } else if (principal != basePrincipal && principal != getCurrentCallerPrincipal()) {
```

Comment:
Caller principal shouldn't be same as a principal in the current security context or souldn't be equals?

Reply:
I think it's safe. We're looking at the exact same principal instance, which indicates that the security context has already been setup.

After authentication (the Jakarta Authentication path) both the request and the security context are set using the exact same principal (the one that was set in the `Subject` by the `ServerAuthModule`. This is exactly the case where we don't want to setup the security context again.